File: server-node/src/queries/get-games.ts
import { Request, Response } from "express"
import { AsyncResult, isFailure, Result } from "../lib/result"
import { Game, makeGameDTO } from "../domain/game"
import { z, ZodError } from "zod"
import { Context } from "../middleware/context"

const GetGamesRequestSchema = z.object({
  playerId: z.string(),
})

export const handle_GetGames = async (
  req: Request,
  res: Response<Game[] | ZodError | "DB_ERR_GET_GAMES_FOR_USER_ID">
) => {
  const parsedBody = GetGamesRequestSchema.safeParse(req.body)
  if (!parsedBody.success) {
    return res.status(400).json(parsedBody.error)
  }
  const { playerId } = parsedBody.data
  const gamesResult = await query_GetGamesForPlayerId(playerId, req.context)
  if (isFailure(gamesResult)) {
    return res.status(500).json(gamesResult.message)
  }

  return res.status(200).json(gamesResult.data)
}

export const query_GetGamesForPlayerId = async (
  playerId: string,
  { Loader: { GameLoader } }: Context
): AsyncResult<Game[], "DB_ERR_GET_GAMES_FOR_USER_ID"> => {
  const gamesResult = await GameLoader.getGamesForPlayerId(playerId)
  if (isFailure(gamesResult)) {
    return gamesResult
  }
  return Result.Success(gamesResult.data.map(makeGameDTO))
}

File: server-python/server/queries/get_games.py
from flask import jsonify, make_response, request
from pydantic import BaseModel

from server.domain.game import Game
from server.models.game import GameLoader, make_game_dto


def query_get_games_for_player_id(playerId: str) -> list[Game]:
    games = GameLoader().get_games_for_player_id(playerId)
    return [make_game_dto(g) for g in games]


class Payload(BaseModel):
    playerId: str


def handle_get_games():
    try:
        payload = Payload(**request.json)
    except Exception:
        return make_response(jsonify({"message": "Invalid payload"}), 400)

    return make_response(jsonify(query_get_games_for_player_id(payload.playerId)), 200)
